### PR TITLE
Ardupilotmega: add OBSTACLE_DISTANCE_3D message

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This is a WIP.
This PR adds the ability to send obstacle distance message, in body frame from a 3-D depth camera. 
We assume the field of view is divided in a 3x3 grid like such:

vlcsnap-2020-08-20-11h18m09s677

We take the closest distance in each individual box, pack it into a NEU frame distance vector in meters and send it.
Furthermore, this message also includes an "ID" field for future implementation of detecting obstacles and tagging them with a  unique ID.

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -5319,6 +5319,17 @@
       <field type="uint32_t" name="active" display="bitmask">Active outputs</field>
       <field type="float[32]" name="actuator">Servo / motor output array values. Zero values indicate unused channels.</field>
     </message>
+    <message id="376" name="VISO_OBSTACLE_DISTANCE">
+      <description>Obstacle distances in front of the 3D depth camera, starting from the top left to bottom right divided in a 3X3 grid. This message assumes NEU frame of reference</description>
+      <field type="uint64_t" name="time_usec" units="us">Timestamp (UNIX Epoch time or time since system boot). The receiving end can infer timestamp format (since 1.1.1970 or since system boot) by checking for the magnitude of the number.</field>
+      <field type="uint8_t" name="sensor_type" enum="MAV_DISTANCE_SENSOR">Class id of the distance sensor type.</field>
+      <field type="uint16_t[9]" name="obstacle_ID"> Unique ID given to each obstacle so that its movement can be tracked. Use 0 if object ID is unknown.</field>
```
Comment (rmackay9) [inline message_definitions/v1.0/common.xml:5326]: normally in mavlink land it seems UINT16_MAX for unknown ids.

Comment (rmackay9) [reply message_definitions/v1.0/common.xml:5326]: we should probably add a "frame" field similar to other messages with a frame field.  AP will only implement the body frame for now but someone will bring this up I'm sure.

Comment (rmackay9) [reply message_definitions/v1.0/common.xml:5326]: Maybe we should call the message "VISION_OBSTACLE_DISTANCE"?

Comment (auturgy) [reply message_definitions/v1.0/common.xml:5326]: Frame, Field of View (x/y) and Orientation (direction/offset of centre of FoV) should be definable (fields) I think.
Suggest name OBSTACLE_GRID_3X3

**Comment by amilcarlucas** [comment]
Please add this upstream in the mavlink/mavlink repo

**Comment by rishabsingh3003** [comment]
Some improvements:
- Removed the message from common, added it to ardupilotmega.xml
- Instead of making this message specifically for a 3*3 box, I decided to make it more generic.. So now we can send any {x,y,z} of an obstacle in body frame and ArduPilot will be able to decide how to sort it based on the body frame position. This makes this message great for other applications like tagging an obstacle with ArucoMarker 
 - Other stuff that are reviewed above

**Comment by auturgy** [comment]
Can you change the PR title to ardupilotmega: add VISION_OBSTACLE_DISTANCE please.

File: message_definitions/v1.0/ardupilotmega.xml [hunk 2]
```diff
@@ -1763,5 +1763,19 @@
       <field type="float" name="max_value">OSD parameter maximum value.</field>
       <field type="float" name="increment">OSD parameter increment.</field>
     </message>
+    <message id="11037" name="VISION_OBSTACLE_DISTANCE">
+      <wip/>
+      <!-- This message is work-in-progress and it can therefore change, and should NOT be used in stable production environments -->
+      <description>Obstacle located as a 3D vector. Maximum of 9 obstacles can be sent in one message.</description>
+      <field type="uint64_t" name="time_usec" units="us">Timestamp (UNIX Epoch time or time since system boot). The receiving end can infer timestamp format (since 1.1.1970 or since system boot) by checking for the magnitude of the number.</field>
```
Comment (tridge) [inline message_definitions/v1.0/ardupilotmega.xml:1770]: i'd prefer ms

File: message_definitions/v1.0/ardupilotmega.xml [hunk 3]
```diff
@@ -1763,5 +1763,19 @@
...
+      <field type="uint8_t" name="sensor_type" enum="MAV_DISTANCE_SENSOR">Class id of the distance sensor type.</field>
+      <field type="uint8_t" name="frame" enum="MAV_FRAME">Coordinate frame of reference.</field>
+      <field type="uint16_t[9]" name="obstacle_ID"> Unique ID given to each obstacle so that its movement can be tracked. Use UINT16_MAX if object ID is unknown or cannot be determined.</field>
```
Comment (tridge) [inline message_definitions/v1.0/ardupilotmega.xml:1773]: i'd prefer a single value, and send multiple msgs with same timestamp

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
to keep packets small, maybe use uint16_t cm?

File: message_definitions/v1.0/ardupilotmega.xml [hunk 4]
```diff
@@ -1763,5 +1763,19 @@
       <field type="float" name="max_value">OSD parameter maximum value.</field>
       <field type="float" name="increment">OSD parameter increment.</field>
     </message>
+    <message id="11037" name="VISION_OBSTACLE_DISTANCE">
```
Comment (tridge) [inline message_definitions/v1.0/ardupilotmega.xml:1766]: rename to OBSTACLE_DISTANCE_3D

**Comment by rmackay9** [comment]
Suggestions from the dev call:

- rename obstacle_distance_3d
- change to a single obstacle per message
- change timestamp to uint32 in milliseconds
- can we remove the max and min distance

**Comment by rishabsingh3003** [comment]
Thanks for the detailed review!
I have taken all the suggestion except: removing the min,max distances. Mostly because, we have no parameter in Proximity lib that checks it currently, and while my script is under development (that uses this new message), I would like a final check at ardupilot so that the vehicle does not respond to garbage or FLOAT_MAX values by mistake. 

Thanks :)

**Comment by rmackay9** [comment]
Merged, thanks.  Let's add the min and max to the proximity library and then remove the fields from this message to reduce bandwidth requirements.

**Comment by rishabsingh3003** [comment]
Thanks for the merge! I have also taken note of the new parameters we need to add :)

**Pull request merged**